runtime/common/src/parachains.rs

```diff
@@ -304,6 +532,46 @@ decl_module! {
 			Ok(())
 		}
 
+		/// Provide a proof that some validator has commited a double-vote.
+		///
+		/// The weight is 0; in order to avoid DoS a `SignedExtension` validation
+		/// is implemented.
+		#[weight = SimpleDispatchInfo::FixedNormal(0)]
+		pub fn report_double_vote(
+			origin,
+			report: DoubleVoteReport,
+			key_owner_proof: Vec<u8>,
+			id: Vec<u8>,
+		) -> DispatchResult {
+			let reporter = ensure_signed(origin)?;
+
+			let validators = <session::Module<T>>::validators();
+			let validator_set_count = validators.len() as u32;
+
+			let key_owner_proof: Proof = Decode::decode(&mut &key_owner_proof[..])
+				.map_err(|_| "Key owner proof decoding failed.")?;
+
+			let session = Default::default(); // TODO: this has to be key_owner_proof.session();
+
+			let id = T::HandleDoubleVote::check_report(
```

Comment:
The logic happening here should happen in `DoubleVoteReport::verify` IMO. We don't want reports about nonexistent validators to make their way into the chain.